File: scripts/figures/fig_5/0_top_validation_FlaiMapper_miRBase/validation_FlaiMapper_miRBase_SRP006788.R
experiments = c('SRR207111_HeLa18-30', 'SRR207112_HeLa18-30_RRP40', 'SRR207113_HeLa18-30_AGO1_2', 'SRR207114_HeLa18-30_AGO1_2_RRP40', 'SRR207115_HeLa18-30_XRN1_2', 'SRR207116_HeLa18-30_N')

for(experiment in experiments) {
  f1 = paste("validation_FlaiMapper_miRBase_SRP006788_",experiment,"_sensitivity.txt",sep="")
  f2 = paste("validation_FlaiMapper_miRBase_SRP006788_",experiment,"_offset.txt",sep="")
  
  f3 = paste("../../../../output/figures/fig_5_extra/validation_FlaiMapper_miRBase_SRP006788_",experiment,"_sensitivity_barplot.svg",sep="")
  f4 = paste("../../../../output/figures/fig_5_extra/validation_FlaiMapper_miRBase_SRP006788_",experiment,"_sensitivity_piechart.svg",sep="")
  
  f5 = paste("../../../../output/figures/fig_5_extra/validation_FlaiMapper_miRBase_SRP006788_",experiment,"_offset_with_miRBase_start-positions.svg",sep="")
  f6 = paste("../../../../output/figures/fig_5_extra/validation_FlaiMapper_miRBase_SRP006788_",experiment,"_offset_with_miRBase_stop-positions.svg",sep="")
  
  
  table_overlap <- read.table(f1,header=T,row.names=1)
  table_error <- read.table(f2,header=T,row.names=1)
  
  # Overlap plot
  svg(f3)
  column <- 3
  x_axes = barplot(table_overlap[,column],ylim=c(0,650),names.arg=c("predicted","not predicted\nsupporting reads","not predicted\nno supporting reads"),ylab="Annotated miRNAs in miRBase 20")
  percentages <- round(table_overlap[,column]/sum(table_overlap[,column])*100,1)
  texting <- paste(table_overlap[,column],rep("\n(",3),percentages,rep("%)",3),sep="")
  text(x_axes,table_overlap[,column]+36,texting)
  dev.off()
  
  
  svg(f4)
  column <- 3
  pie(table_overlap[,column],labels=paste(c("predicted","not predicted","no reads"),texting,sep=": "),col=c("darkgreen","red","gray"),main="miRBase predicted by FlaiMapper\nsample SRP006788")
  dev.off()
  
  
  # Error plot 5p
  svg(f5)
  column <- 5
  percentages <- round(table_error[,column]/sum(table_error[,column])*100,1)
  x_axes = barplot(percentages,ylim=c(0,110),names.arg=row.names(table_error),
  main="5'-error between miRBase miRNAs and FlaiMapper",ylab="% of occurance",xlab="Difference in 5' position between predicted fragment and annotated miRNA")
  texting <- paste(table_error[,column],rep("\n",21),percentages,rep("%",12),sep="")
  text(x_axes,percentages+8,texting)
  dev.off()
  
  
  # Error plot 3p
  svg(f6)
  column <- 6
  percentages <- round(table_error[,column]/sum(table_error[,column])*100,1)
  x_axes = barplot(percentages,ylim=c(0,110),names.arg=row.names(table_error)
  ,main="3'-error between miRBase miRNAs and FlaiMapper",ylab="% of occurance",xlab="Difference in 3' position between predicted fragment and annotated miRNA")
  texting <- paste(table_error[,column],rep("\n",21),percentages,rep("%",12),sep="")
  text(x_axes,percentages+8,texting)
  dev.off()
  
}

File: scripts/figures/fig_5/0_top_validation_FlaiMapper_miRBase/validation_FlaiMapper_miRBase_SRP006788.py
#!/usr/bin/env python

import sys
sys.path.append("../../../../src")



from miRBase import *

from FragmentContainer import FragmentContainer
from FragmentFinder import FragmentFinder
from AlignmentParser import AlignmentParser
from AlignmentDirectory import AlignmentDirectory
from FlaiMapperObject import FlaiMapperObject



verbosity = "quiet"
miRNAs = miRBase("../../../../share/annotations/miRBase_20/miRNA.dat")
dataset_id = "SRP006788"




experiments = ['SRR207111_HeLa18-30', 'SRR207112_HeLa18-30_RRP40', 'SRR207113_HeLa18-30_AGO1_2', 'SRR207114_HeLa18-30_AGO1_2_RRP40', 'SRR207115_HeLa18-30_XRN1_2', 'SRR207116_HeLa18-30_N']
for experiment in experiments:
	alignments = []
	alignments.append(AlignmentDirectory("../../../../share/small_RNA-seq_alignments/"+dataset_id+"/"+experiment,verbosity))
	
	ncrna_library_names = {}
	for alignment in alignments:
		for name in alignment.index:
			ncrna_library_names[name] = True
	ncrna_library_names = ncrna_library_names.keys()
	
	i = 0
	
	# Crosslink miRBase with reference ncRNAs
	links = {}
	for miRNA in miRNAs.get_miRNAs():
		for name in ncrna_library_names:
			if(name.lower().find("mir") > -1):
				flaimapper_name_raw = name.split("HUGO-Symbol=")[1].split("&")[0]
				
				if(flaimapper_name_raw in miRNA.get_parameter("aliases")):
					links[name] = miRNA.params["name"]
					
					i += 1
	
	# Load flaimapper
	flaimapper = FlaiMapperObject(verbosity)
	for alignment in alignments:
		flaimapper.add_alignment_directory(alignment)
	results = flaimapper.count_reads_per_region(miRNAs,links,10)
	
	
	keys = sorted(results.keys())
	keys.append("stacked")
	
	# Export found-backs
	fh = open("validation_FlaiMapper_miRBase_"+dataset_id+"_"+experiment+"_sensitivity.txt","w")
	
	factors = ["predicted","not_predicted_with_reads","not_predicted_no_reads"]
	fh.write("factor\t"+"\t".join(keys)+"\n")
	for factor in factors:
		line = factor
		line += "\t"+str(results[keys[0]][factor])
		line += "\t"+str(results[keys[1]][factor])
		line += "\t"+str(results[keys[0]][factor]+results[keys[1]][factor])
		fh.write(line+"\n")
	
	fh.close()
	
	
	# Export Errors
	fh = open("validation_FlaiMapper_miRBase_"+dataset_id+"_"+experiment+"_offset.txt","w")
	
	header = "error"
	for k in keys:
		for prime in ["_5p","_3p"]:
			header += "\t"+k+prime
	
	fh.write(header+"\n")
	xkeys = ["<-5",-5,-4,-3,-2,-1,0,1,2,3,4,5,">5"]
	
	for n in xkeys:
		line = str(n)
		
		for k in keys:
			for prime in ["_5p","_3p"]:
				if(k != "stacked"):
					line += "\t"+str(results[k]["error"+prime][n])
				else:
					line += "\t"+str(results[keys[0]]["error"+prime][n]+results[keys[1]]["error"+prime][n])
		
		fh.write(line+"\n")

	fh.close()


